libuavcan/src/transport/can_acceptance_filter_configurator.cpp

```diff
@@ -0,0 +1,256 @@
+}
+
+int16_t CanAcceptanceFilterConfigurator::applyConfiguration(void)
+{
+    CanFilterConfig filter_conf_array[MaxCanAcceptanceFilters];
+    const unsigned int filter_array_size = multiset_configs_.getSize();
+
+    if (filter_array_size > MaxCanAcceptanceFilters)
+    {
+        UAVCAN_ASSERT(0);
+        return -ErrLogic;
+    }
+
+    for (uint16_t i = 0; i < filter_array_size; i++)
+    {
+        CanFilterConfig temp_filter_config = *multiset_configs_.getByIndex(i);
+        if (multiset_configs_.getByIndex(i) == NULL)
+        {
+            UAVCAN_ASSERT(0);
+        }
+
+        if ((temp_filter_config.id != 0) || (temp_filter_config.mask != 0))
+        {
+            filter_conf_array[i] = temp_filter_config;
+        }
+    }
+
+    ICanDriver& can_driver = node_.getDispatcher().getCanIOManager().getCanDriver();
+    for (uint8_t i = 0; i < node_.getDispatcher().getCanIOManager().getNumIfaces(); i++)
+    {
+        ICanIface* iface = can_driver.getIface(i);
+        if (iface == NULL)
+        {
+            UAVCAN_ASSERT(0);
+            break;
+        }
+        int16_t num = iface->configureFilters(reinterpret_cast<CanFilterConfig*>(&filter_conf_array),
+                                              static_cast<uint16_t>(filter_array_size));
+        if (num < 0)
+        {
+            return -ErrDriver;
+        }
+    }
+
+    return 0;
+}
+
+int CanAcceptanceFilterConfigurator::configureFilters()
+{
+    if (getNumFilters() == 0)
+    {
+        UAVCAN_TRACE("CanAcceptanceFilter", "No HW filters available");
+        return -ErrDriver;
+    }
+
+    int16_t fill_array_error = fillArray();
+    if (fill_array_error != 0)
+    {
+        UAVCAN_TRACE("CanAcceptanceFilter::fillArray", "Failed to execute fillArray()");
+        return fill_array_error;
+    }
+
+    int16_t compute_configuration_error = computeConfiguration();
+    if (compute_configuration_error != 0)
+    {
+        UAVCAN_TRACE("CanAcceptanceFilter", "Failed to compute optimal acceptance fliter's configuration");
+        return compute_configuration_error;
+    }
+
+    if (applyConfiguration() != 0)
+    {
+        UAVCAN_TRACE("CanAcceptanceFilter", "Failed to apply HW filter configuration");
+        return -ErrDriver;
+    }
+
+    return 0;
+}
+
+uint16_t CanAcceptanceFilterConfigurator::getNumFilters() const
+{
+
+    static const uint16_t InvalidOut = 0xFFFF;
+    uint16_t out = InvalidOut;
+    ICanDriver& can_driver = node_.getDispatcher().getCanIOManager().getCanDriver();
+
+    for (uint8_t i = 0; i < node_.getDispatcher().getCanIOManager().getNumIfaces(); i++)
+    {
+        const ICanIface* iface = can_driver.getIface(i);
+        if (iface == NULL)
+        {
+            UAVCAN_ASSERT(0);
+            out = 0;
+            break;
+        }
+        const uint16_t num = iface->getNumFilters();
+        out = min(out, num);
+        if (num > MaxCanAcceptanceFilters)
+        {
+            return MaxCanAcceptanceFilters;
+        }
```

Comment:
Ha, a bug!

The objective is to find the lowest number of HW filters. The function is supposed to traverse all interfaces and find the lowest value.

The condition here breaks the logic if one of the interfaces has more filters than `MaxCanAcceptanceFilters`. For example, if there are three interfaces with 5, 6, and 7 filters, and `MaxCanAcceptanceFilters` is 7, the function will return 7 instead of 5.